// watson-developer-cloud/visual-recognition/v3.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var extend = require("extend");
var GeneratedVisualRecognitionV3 = require("./v3-generated");
var VisualRecognitionV3 = /** @class */ (function (_super) {
    __extends(VisualRecognitionV3, _super);
    function VisualRecognitionV3(options) {
        var _this = this;
        // For backward compatibility, allow version to be passed in version_date.
        var _options = extend({}, options);
        _options.version = _options.version_date || _options.version;
        _this = _super.call(this, _options) || this;
        if (_options.api_key) {
            console.warn('WARNING: Visual Recognition instances are no longer accepting `api_key` for authentication. Use `iam_apikey`.');
        }
        return _this;
    }
    VisualRecognitionV3.prototype.classify = function (params, callback) {
        if (params && params.image_file) {
            params.images_file = params.image_file;
        }
        var newParams = params ? extend.apply(void 0, [{}, params].concat((params.parameters || {}))) : params;
        return _super.prototype.classify.call(this, newParams, callback);
    };
    VisualRecognitionV3.prototype.detectFaces = function (params, callback) {
        if (params && params.image_file) {
            params.images_file = params.image_file;
        }
        var newParams = params ? extend.apply(void 0, [{}, params].concat((params.parameters || {}))) : params;
        return _super.prototype.detectFaces.call(this, newParams, callback);
    };
    VisualRecognitionV3.prototype.retrainClassifier = function (params, callback) {
        console.warn("WARNING: retrainClassifier() was renamed to updateClassifier(). Support for retrainClassifier() will be removed in the next major release");
        return _super.prototype.updateClassifier.call(this, params, callback);
    };
    VisualRecognitionV3.prototype.recognizeText = function (params, callback) {
        console.warn(VisualRecognitionV3.betaError);
    };
    VisualRecognitionV3.prototype.createCollection = function (params, callback) {
        console.warn(VisualRecognitionV3.betaError);
    };
    VisualRecognitionV3.prototype.getCollection = function (params, callback) {
        console.warn(VisualRecognitionV3.betaError);
    };
    VisualRecognitionV3.prototype.listCollections = function (params, callback) {
        console.warn(VisualRecognitionV3.betaError);
    };
    VisualRecognitionV3.prototype.deleteCollection = function (params, callback) {
        console.warn(VisualRecognitionV3.betaError);
    };
    VisualRecognitionV3.prototype.addImage = function (params, callback) {
        console.warn(VisualRecognitionV3.betaError);
    };
    VisualRecognitionV3.prototype.listImages = function (params, callback) {
        console.warn(VisualRecognitionV3.betaError);
    };
    VisualRecognitionV3.prototype.getImage = function (params, callback) {
        console.warn(VisualRecognitionV3.betaError);
    };
    VisualRecognitionV3.prototype.deleteImage = function (params, callback) {
        console.warn(VisualRecognitionV3.betaError);
    };
    VisualRecognitionV3.prototype.setImageData = function (params, callback) {
        console.warn(VisualRecognitionV3.betaError);
    };
    VisualRecognitionV3.prototype.getImageData = function (params, callback) {
        console.warn(VisualRecognitionV3.betaError);
    };
    VisualRecognitionV3.prototype.deleteImageData = function (params, callback) {
        console.warn(VisualRecognitionV3.betaError);
    };
    VisualRecognitionV3.prototype.findSimilar = function (params, callback) {
        console.warn(VisualRecognitionV3.betaError);
    };
    VisualRecognitionV3.VERSION_DATE_2016_05_20 = '2016-05-20';
    VisualRecognitionV3.betaError = new Error('As of September 8, 2017, the beta period for Similarity Search is closed.' +
        'For more information, see [Visual Recognition API – Similarity Search Update]' +
        '(https://www.ibm.com/blogs/bluemix/2017/08/visual-recognition-api-similarity-search-update)');
    return VisualRecognitionV3;
}(GeneratedVisualRecognitionV3));
module.exports = VisualRecognitionV3;
